fix(AI Agent Node): Fix tools agent when using memory and Anthropic models

OlegIvaniv opened the pull request:
## Summary
When using Anthropic models paired with memory, we discovered an issue with a Tools agent which would throw an error when populating the memory. To fix it, we need to check`AgentFinish` step if it contains string or array output(in Anthropic's case) and merge the text-only array of outputs to a single item.

## Related Linear tickets, Github issues, and Community forum posts
- https://community.n8n.io/t/the-memory-is-not-working-with-the-ai-agent-tools-agent-node/51899
## Review / Merge checklist

- [ ] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

OlegIvaniv committed:
fix(AI Agent): Fix tools agent when using memory and Anthropic models

netroy commented on packages/@n8n/nodes-langchain/package.json:
```diff
@@ -137,17 +137,17 @@
-    "@langchain/ollama": "^0.0.2",
-    "@langchain/openai": "0.2.5",
-    "@langchain/pinecone": "0.0.8",
+    "@langchain/groq": "0.0.17",
+    "@langchain/mistralai": "0.0.28",
+    "@langchain/community": "0.2.28",
+    "@langchain/core": "0.2.27",
+    "@langchain/google-genai": "0.0.26",
+    "@langchain/google-vertexai": "0.0.26",
```
can we revert this sorting to be alphabetical again, to make the diff easier to read 🙏🏽

OlegIvaniv replied:
Good catch, done!

OlegIvaniv committed:
Add back step output parsing

OlegIvaniv committed:
Revert ordering of modules

burivuhster commented on packages/@n8n/nodes-langchain/nodes/agents/Agent/agents/ToolsAgent/execute.ts:
```diff
@@ -93,7 +93,43 @@ export async function toolsAgentExecute(this: IExecuteFunctions): Promise<INodeE
+		type AgentMultiOutputFinish = AgentFinish & {
+			returnValues: { output: Array<{ text: string; type: string; index: number }> };
+		};
+		const agentFinishSteps = steps as AgentMultiOutputFinish | AgentFinish;
```
Should we make a copy here? Modifying function arguments may potentially lead to some hard-to-trace issues in the future.

OlegIvaniv replied:
Cloning this object using `structuredClone` doesn't seem to work as it removes some class bindings Langchain requires so sticking with overriding the step content

burivuhster commented on packages/@n8n/nodes-langchain/nodes/agents/Agent/agents/ToolsAgent/execute.ts:
```diff
@@ -93,7 +93,43 @@ export async function toolsAgentExecute(this: IExecuteFunctions): Promise<INodeE
 	const outputParser = (await getOptionalOutputParsers(this))?.[0];
 	let structuredOutputParserTool: DynamicStructuredTool | undefined;
+	/**
+	 * Handles the agent text output and transforms it in case of multi-output.
```
Can you please add some description why do we need this? Or examples of cases when this will be used.

OlegIvaniv replied:
Added

OlegIvaniv committed:
Merge branch 'master' into ai-259-issue-with-anthropic-model-tools-agent-memory

OlegIvaniv committed:
Clone steps

OlegIvaniv committed:
Address review comments

OlegIvaniv committed:
Add json formatting for the comment

OlegIvaniv commented:
@burivuhster I've added a more detailed comment explaining what the `handleAgentFinishOutput` does. I also removed the langchain version updates and will do that in a tech-debt PR in conjunction with moving the versions to a catalog. Can you have one more look, please?

burivuhster commented on packages/@n8n/nodes-langchain/nodes/agents/Agent/agents/ToolsAgent/execute.ts:
```diff
@@ -96,6 +96,59 @@ export async function toolsAgentExecute(this: IExecuteFunctions): Promise<INodeE
 	/**
 	 * Handles the agent text output and transforms it in case of multi-output.
 	 *
+	 * This method is necessary to handle different output formats from various language models.
+	 * Specifically, it checks if the agent step is the final step (contains returnValues) and determines
+	 * if the output is a simple string (e.g., from OpenAI models) or an array of outputs (e.g., from Anthropic models).
+	 *
+	 * If the output is an array of text outputs, this method will concatenate them into a single string,
+	 * ensuring compatibility with downstream processes that expect a single output string.
+	 *
+	 * Examples:
+	 * 1. Anthropic model output:
+	 *    {
+	 *      "output": [
+	 *        {
+	 *          "index": 0,
+	 *          "type": "text",
+	 *          "text": "The result of the calculation is approximately 1001.8166..."
+	 *        }
+	 *      ]
+	 *    }
+	 *
+	 * 2. OpenAI model output:
+	 *    {
+	 *      "output": "The result of the calculation is approximately 1001.82..."
+	 *    }
+	 *
+	 * This method ensures consistent handling of outputs regardless of the model used,
+	 * providing a unified output format for further processing.
+	 *
```
👍👍👍

burivuhster commented on packages/@n8n/nodes-langchain/nodes/agents/Agent/agents/ToolsAgent/execute.ts:
````diff
@@ -96,6 +96,59 @@ export async function toolsAgentExecute(this: IExecuteFunctions): Promise<INodeE
+	 * This method ensures consistent handling of outputs regardless of the model used,
+	 * providing a unified output format for further processing.
+	 *
+	 * This method is necessary to handle different output formats from various language models.
+	 * Specifically, it checks if the agent step is the final step (contains returnValues) and determines
+	 * if the output is a simple string (e.g., from OpenAI models) or an array of outputs (e.g., from Anthropic models).
+	 *
+	 * Examples:
+	 * 1. Anthropic model output:
+	 * ```json
+	 *    {
+	 *      "output": [
+	 *        {
+	 *          "index": 0,
+	 *          "type": "text",
+	 *          "text": "The result of the calculation is approximately 1001.8166..."
+	 *        }
+	 *      ]
+	 *    }
+	 *```
+	 * 2. OpenAI model output:
+	 * ```json
+	 *    {
+	 *      "output": "The result of the calculation is approximately 1001.82..."
+	 *    }
+	 * ```
+	 *
+	 * This method ensures consistent handling of outputs regardless of the model used.
+	 *
````
It seems like part of the comment was duplicated

OlegIvaniv replied:
Oops, my bad. Should be fixed

burivuhster reviewed (commented):
Kudos for a detailed comment! Please check the duplication, otherwise good to go.

OlegIvaniv committed:
Fix duplicate comment

burivuhster reviewed (approved):
🚀

janober commented:
Got released with [`n8n@1.58.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.58.0)
